### PR TITLE
Refactor: prevent subscription renewals with gateway transaction IDs already used previously

### PR DESCRIPTION
**Pull request by glaubersilva**
Resolves [GIVE-1078]

## Description

This PR implements a defensive approach to prevent duplicated renewal donations in cases where the `SubscriptionRenewalDonationCreated` class is used in the wrong context.

For example:

1. The donor creates a new subscription and it gets the gateway transaction ID for the first transaction immediately;
2. Even so the gateway sends a webhook notification related to this first transaction;
3. We receive this webhook notification and we don't check if the notification is related to the first transaction or if it's a renewal;
4. The `SubscriptionRenewalDonationCreated` class is used to handle the first transaction instead of the `SubscriptionFirstDonationCompleted` class;
5. As a result a new transaction is created with the same gateway transaction ID from the first donation, in other words, the subscription duplicates the first payment.

## Affects

Gateways using the webhooks event handler classes for subscriptions

## Visuals

image

## Testing Instructions

In you console, run the following command: `composer test -- --filter SubscriptionRenewalDonationCreatedTest`

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

[GIVE-1078]: https://stellarwp.atlassian.net/browse/GIVE-1078?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php [hunk 1]
```diff
@@ -24,7 +24,13 @@ public function __invoke(
     ) {
         $subscription = give()->subscriptions->getByGatewaySubscriptionId($gatewaySubscriptionId);
 
-        if ( ! $subscription) {
+        if ( ! $subscription || $subscription->initialDonation()->gatewayTransactionId === $gatewayTransactionId) {
+            return;
+        }
```
Comment (JasonTheAdams) [inline src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php:29]: Might be worth separating these circumstances and adding some logging, so customers/support can see this happening.

Comment (jonwaldstein) [reply src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php:29]: I agree, we should add some logging here

File: tests/Unit/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreatedTest.php [hunk 2]
```diff
@@ -44,4 +44,52 @@ public function testShouldCreateRenewalDonation()
...
+    {
+        $subscription = Subscription::factory()->createWithDonation();
+
+        $duplicatedGatewayTransactionId = 'duplicated-gateway-transaction-id';
+
+        // #1 Renewal Donation
+        give(SubscriptionRenewalDonationCreated::class)($subscription->gatewaySubscriptionId,
+            $duplicatedGatewayTransactionId);
+
+        // #2 Renewal Donation - This one should not be created
```
Comment (jonwaldstein) [inline tests/Unit/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreatedTest.php:87]: Love these! 🦾

**Review by jonwaldstein — APPROVED** [review]
@glaubersilva great job! as mentioned, some additional logging would be helpful - however I know this was a Fun Friday endeavor and your contribution is much appreciated ✊

File: src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php (hunk 1 above)
Comment (glaubersilva) [reply src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php]: I agree, these logs can be really helpful. Resolved in 156a0f280a2f1a45c3403954d3e09a935408fab0 and bd025e3bab29a8bb49124b979568107db2420cf1

File: src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php [hunk 3]
```diff
@@ -25,6 +26,48 @@ public function __invoke(
...
+
+        if ($subscription->initialDonation()->gatewayTransactionId === $gatewayTransactionId) {
+            PaymentGatewayLog::error(
+                sprintf('The renewal was not created for the gateway transaction ID %s because the initial donation of the subscription %s is already using the informed gateway transaction ID %s.',
+                    $gatewayTransactionId, $subscription->id, $gatewaySubscriptionId),
+                [
+                    'Gateway Subscription ID' => $gatewaySubscriptionId,
+                    'Gateway Transaction ID' => $gatewayTransactionId,
+                    'Message' => $message,
+                    'Subscription' => $subscription,
```
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php]: Got it! Resolved in d53629ef7cd33717f89e3afe409fa6a5d1b933c0

File: src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php [hunk 4]
```diff
@@ -25,6 +26,48 @@ public function __invoke(
...
+
+        if ($donation) {
+            PaymentGatewayLog::error(
+                sprintf('The renewal was not created for the gateway transaction ID %s because the donation %s is already using the informed gateway transaction ID %s.',
+                    $gatewayTransactionId, $donation->id, $gatewaySubscriptionId),
+                [
+                    'Gateway Subscription ID' => $gatewaySubscriptionId,
+                    'Gateway Transaction ID' => $gatewayTransactionId,
+                    'Message' => $message,
+                    'Donation' => $donation,
```
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/Framework/PaymentGateways/Webhooks/EventHandlers/SubscriptionRenewalDonationCreated.php]: Got it! Resolved in d53629ef7cd33717f89e3afe409fa6a5d1b933c0

**Review by rickalday — APPROVED** [review]
Passed manual QA tests.

**Pull request merged**